### ISSUE TITLE
Question / Comment: File size increasing substantially with links

### ISSUE THREAD

**Issue by IanMelbourne93** [pymupdf/PyMuPDF#480]
Hi guys,
I am currently adding an image of the company logo to the corner of each page of my PDF, I am then adding a link to the first page (which contains a table of contents with its own links). My issue is that these links appear to substantially increase the size of my PDF. For a 100 page PDF, the size goes from 46MB to about 63MB with the corner image and links (image is a 5kB embedded vector pdf embedded using showPDFpage ). I didn't previously see this when using older versions of PyMuPDF, namely when reuse xref was available in showPDFpage function. 
Am wondering if there is anything obvious I am doing wrong? Each page should link to the same place so I am happy to reuse any Xref that I can.

My function to add the logo is below. Note that I have needed to do a few workarounds to remove the black border and to ensure that the first page is not skipped (I'm not sure why these occur, might be a bug, but my workarounds consistently work):

    def AddLogo(pdfname, logolink=r"C:/logo.pdf", exclude_pages=[0], links=True, linked_page=0):
        try:
            import fitz
        except ImportError:
            print("Unable to import PyMuPDF (fitz), library not found")
            return
        if isinstance(exclude_pages, int):
            exclude_pages = [exclude_pages]
        rect = fitz.Rect(10, 10, 87, 50)
        overlaypdf = fitz.open(logolink)  # open the logo as a PDF
        doc = fitz.open(pdfname)
        doc.saveIncr()
        l = {'kind': 1, "xref": 1, "from": rect, "page": 1, "to": fitz.Point(0, 0), "zoom": 0.0}
        pages = [doc[0]] + [p for p in doc]  # For some reason I need to duplicate the first page...
        for iter, page in enumerate(pages):
            if iter in exclude_pages:
                continue
            # Overlay the logo PDF, pno refers to page num of logo PDF
            xref = page.showPDFpage(rect, overlaypdf, pno=0, overlay=True, keep_proportion=True)
    
            if links:
                new_link = l
                new_link["kind"] = 1
                new_link["page"] = linked_page  # Link to home
                new_link["from"] = rect
                new_link["xref"] = 1
                print(new_link)
                page.insertLink(new_link)
                doc.saveIncr()
                L = page.firstLink
                while (L != None):
                    # Removing the black border from around the link by setting width to 0.
                    border = L.border  # Making sure we grab the border first
                    if L.xref == 0:
                        L = L.next
                        continue
                    if isinstance(border, dict):
                        border['width'] = 0
                        L.setBorder(border)
                    else:
                        L.setBorder(width=0)
                    L = L.next
            doc.saveIncr()
        for page in doc:
            L = page.firstLink
            while (L != None):
                # Removing the black border from around the link by setting width to 0.
                border = L.border  # Making sure we grab the border first
                if L.xref == 0:
                    L = L.next
                    continue
                if isinstance(border, dict):
                    border['width'] = 0
                    L.setBorder(border)
                else:
                    L.setBorder(width=0)
                L = L.next
            doc.saveIncr()
        doc.close()

**Comment by JorjMcKie** [comment]
I haven't reviewed your code in detail, but as a first observation:
You are doing lots of ``saveIncr()``. This will cause physical increases of the PDF file each time you issue them.
And I do not see why they are ever necessary  - except for exactly one time: immediately before the close.
As long as you change things in the PDF, that will happen in memory only. Upon save, all accumulated changes will be written out.
Is there any special reason why you must use ``saveIncr()``? This will **always** increase filesize of course. Saving to a new file in contrast offers compression and garbage collection options.

BTW: The xref parameter of ``showPDFpage`` is no longer required because I now store the xref somewhere for reuse. So the space saving feature of it is still there.

**Comment by IanMelbourne93** [comment]
Hey Jorj, thanks for the reply.
The saveIncr() were honestly there for debugging purposes as I was having trouble removing the black border around the pdf that popped up when I added the link to the image. Not sure why I had such trouble and I don't think I got to the bottom of it, but the code I was using previously didn't work and it took a good day to figure it out.
I have rerun the code without the saveincr() calls and low and behold, that solved my issue. The filesize is now 46MB. Thank you very much for that, it has been bugging me for a while.

In fact, removing the saveIncr() seem to have solved my other issue that I was seeing where I needed to duplicate the first page in that loop in order for my indexing to work.

Also thank you on the library in general, one of my favorites for this kind of thing, so handy.
